connection_pool/round_robin.go

```diff
@@ -60,58 +61,40 @@ func (r *RoundRobinStrategy) IsEmpty() bool {
 	return r.size == 0
 }
 
-func (r *RoundRobinStrategy) CloseConns() []error {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
-	errs := make([]error, len(r.conns))
-
-	for i, conn := range r.conns {
-		errs[i] = conn.Close()
-	}
-
-	return errs
-}
-
 func (r *RoundRobinStrategy) GetNextConnection() *tarantool.Connection {
 	r.mutex.RLock()
 	defer r.mutex.RUnlock()
 
-	// We want to iterate through the elements in a circular order
-	// so the first element in cycle is connections[next]
-	// and the last one is connections[next + length].
-	next := r.nextIndex()
-	cycleLen := len(r.conns) + next
-	for i := next; i < cycleLen; i++ {
-		idx := i % len(r.conns)
-		if r.conns[idx].ConnectedNow() {
-			if i != next {
-				atomic.StoreUint64(&r.current, uint64(idx))
-			}
-			return r.conns[idx]
-		}
+	if r.size == 0 {
+		return nil
 	}
-
-	return nil
+	return r.conns[r.nextIndex()]
 }
 
 func NewEmptyRoundRobin(size int) *RoundRobinStrategy {
 	return &RoundRobinStrategy{
 		conns:       make([]*tarantool.Connection, 0, size),
-		indexByAddr: make(map[string]int),
+		indexByAddr: make(map[string]uint),
 		size:        0,
+		current:     0,
 	}
 }
 
 func (r *RoundRobinStrategy) AddConn(addr string, conn *tarantool.Connection) {
 	r.mutex.Lock()
 	defer r.mutex.Unlock()
 
-	r.conns = append(r.conns, conn)
-	r.indexByAddr[addr] = r.size
-	r.size += 1
+	if idx, ok := r.indexByAddr[addr]; ok {
+		r.conns[idx] = conn
+	} else {
+		r.conns = append(r.conns, conn)
+		r.indexByAddr[addr] = r.size
+		r.size += 1
+	}
 }
 
-func (r *RoundRobinStrategy) nextIndex() int {
-	return int(atomic.AddUint64(&r.current, uint64(1)) % uint64(len(r.conns)))
+func (r *RoundRobinStrategy) nextIndex() uint {
+	ret := r.current % r.size
+	r.current++
+	return ret
```

Comment:
I guess atomic operation is unnecessary since we have `r.mutex.RLock()` in `GetNextConnection`?

Reply:
Yes, it is.

Reply:
In addition, the counter was used incorrectly if we forget about the mutex. Due to the fact that we increment and set the value, intersections caused by data race are possible:

1.

thread1: GetNextConnection() -> nextIndex() = 11
thread2: GetNextConnection() -> nextIndex() = 12
threadN: GetNextConnection() -> nextIndex() = 1N
thread10: GetNextConnection() -> nextIndex() = 20

2.

thread1: GetNextConnection() -> StoreUint64(12)

Thus, this counter will not give the expected result.

Reply:
Thank you for explanation!